src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs

```diff
@@ -47,66 +47,104 @@ private static void FreeDeconstructionVariables(ArrayBuilder<DeconstructionVaria
-                if (boundRHS.Kind == BoundKind.TupleLiteral)
-                {
-                    // Let's fix the literal up by figuring out its type
-                    // For declarations, that means merging type information from the LHS and RHS
-                    // For assignments, only the LHS side matters since it is necessarily typed
-                    TypeSymbol lhsAsTuple = MakeMergedTupleType(checkedVariables, (BoundTupleLiteral)boundRHS, node, Compilation, diagnostics);
-                    if ((object)lhsAsTuple != null)
-                    {
-                        boundRHS = GenerateConversionForAssignment(lhsAsTuple, boundRHS, diagnostics);
-                    }
-                }
-                else
-                {
-                    Error(diagnostics, ErrorCode.ERR_DeconstructRequiresExpression, right);
-                }
-
-                if ((object)boundRHS.Type == null)
-                {
-                    // we could still not infer a type for the RHS
-                    FailRemainingInferences(checkedVariables, diagnostics);
-
-                    return new BoundDeconstructionAssignmentOperator(
-                                node, isDeclaration, FlattenDeconstructVariables(checkedVariables), boundRHS,
-                                ImmutableArray<BoundDeconstructionDeconstructStep>.Empty, ImmutableArray<BoundDeconstructionAssignmentStep>.Empty,
-                                voidType, hasErrors: true);
-                }
+                // we could still not infer a type for the RHS
+                FailRemainingInferences(checkedVariables, diagnostics);
+
+                return new BoundDeconstructionAssignmentOperator(
+                            node, isDeclaration, FlattenDeconstructVariables(checkedVariables), boundRHS,
+                            ImmutableArray<BoundDeconstructionDeconstructStep>.Empty,
+                            ImmutableArray<BoundDeconstructionAssignmentStep>.Empty,
+                            ImmutableArray<BoundDeconstructionAssignmentStep>.Empty,
+                            ImmutableArray<BoundDeconstructionConstructionStep>.Empty,
+                            GetSpecialType(SpecialType.System_Void, diagnostics, node),
+                            hasErrors: true);
             }
+
             var deconstructionSteps = ArrayBuilder<BoundDeconstructionDeconstructStep>.GetInstance(1);
+            var conversionSteps = ArrayBuilder<BoundDeconstructionAssignmentStep>.GetInstance(1);
             var assignmentSteps = ArrayBuilder<BoundDeconstructionAssignmentStep>.GetInstance(1);
-            bool hasErrors = !DeconstructIntoSteps(new BoundDeconstructValuePlaceholder(boundRHS.Syntax, boundRHS.Type), node, diagnostics, checkedVariables, deconstructionSteps, assignmentSteps);
+            var constructionStepsOpt = isDeclaration ? null : ArrayBuilder<BoundDeconstructionConstructionStep>.GetInstance(1);
+
+            bool hasErrors = !DeconstructIntoSteps(
```

Comment:
>DeconstructIntoSteps [](start = 30, length = 20)

Am I understanding that this takes the expression which is the bound tuple right-hand-side of the resolved type and disassembles the expression to tease out the conversions from the underlying evaluations so they can be reordered?